Enhancement on Substrate Node Template (#8473)
paritytech/substrate: docs/node-template-release.md

# Node Template Release Process

1. It has to be in a github checkout directory with your current work committed into
`https://github.com/paritytech/substrate/`, because this is where the build script will check out

> from. Assume you are in root directory of Substrate. Run:

Reviewer: this is confusing to me, can you rephrase it?

Author: @apopiak this paragraph is updated.

	```bash
	cd .maintain/
	./node-template-release.sh <output tar.gz file>
	```

2. Expand the output tar gzipped file and replace files in current Substrate Node Template
by running the following command.

	```bash
	# This is where the tar.gz file uncompressed
	cd substrate-node-template
	# rsync with force copying. Note the slash at the destination directory is important
	rsync -avh * <destination node-template directory>/
	# For dry-running add `-n` argument
	# rsync -avhn * <destination node-template directory>/
	```

> 	The above command only copy existing files from the source to destination, but not deleting
> 	files/directories that are removed from the source. So manually check and remove them in the
> 	destination.

Reviewer: ```suggestion
	The above command only copies existing files from the source to the destination, but does not delete
	files/directories that are removed from the source. So you need to manually check and remove them in the
	destination.
```

3. There actually are three packages in Node Template, `node-template` (the node), `node-template-runtime`
(the runtime), and `pallet-template`, and each has its own `Cargo.toml`. We need to manually update:

	- Each dependency is listed in expanded form and linked to a certain Substrate commit, such as:

		```toml
		[dev-dependencies.sp-core]
		default-features = false
		git = 'https://github.com/paritytech/substrate.git'
		rev = 'c1fe59d060600a10eebb4ace277af1fee20bad17'
		version = '3.0.0'
		```

> 		We will update each of them to the shortened form and linked to the
> 		[crate registry](https://crates.io/). After confirming the versioned package is published in
> 		crate, the above will become:

Reviewer: ```suggestion
		We will update each of them to the shortened form and linked to the Rust
		[crate registry](https://crates.io/). After confirming the versioned package is published in
		the crate, the above will become:
```

		```toml
		[dev-dependencies]
		sp-core = { version = '3.0.0', default-features = false }
		```

	P.S: This step can be automated if we update `node-template-release` package in

> 	`.maintain/node-template-release`.

Reviewer: this sounds like a todo?

Author: It is. But given there are 626 issues in Substrate repo already, I don't want to add another one that is kind of low priority.

Reviewer: I don't think it makes sense not to create issues and have things in code, but :shrug:

> 4. Once the three `Cargo.toml`s are updated, compile and confirm Node Template builds. Then
> commit the changes to a new branch in [Substrate Node Template](https://github.com/substrate-developer-hub/substrate-node-template), and make a PR.

Reviewer: ```suggestion
4. Once the three `Cargo.toml`s are updated, compile and confirm that the Node Template builds. Then
commit the changes to a new branch in [Substrate Node Template](https://github.com/substrate-developer-hub/substrate-node-template), and make a PR.
```

> 	> Notice there are chances the code in Substrate Node Template works with the linked Substrate git
> 	commit, but not with published packages due to the latest unpublished (yet) features. In this case,
> 	resume that section of Node Template to its previous version to ensure Node Template builds.

Reviewer: ```suggestion
	> Note that there is a chance the code in Substrate Node Template works with the linked Substrate git
	commit but not with published packages due to the latest (as yet) unpublished features. In this case,
	rollback that section of the Node Template to its previous version to ensure the Node Template builds.
```
